### PR TITLE
Allow Coercion between hive Union and hive  struct for Hive ORC table

### PR DESCRIPTION
**Pull request by leetcode-1533**
## Description
Allow Coercion between hive Union and hive struct, at LinkedIn we have the need to declare the table schema as struct and reading some partitions in union definition. 

Since Trino internally treat union as struct, this PR allows coercions between struct and union(after union get converted to Struct)

Related PR:
https://github.com/trinodb/trino/pull/3483
https://github.com/trinodb/trino/commit/e071da402efb4b95f6c8b13b2a241e86e446e2a8

## Non-technical explanation

Allow Coercion between hive Union and hive  struct

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Hive
*Allow Coercion between Union and Struct typed columns.
```

### CODE REVIEW

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestReadUniontype.java [hunk 1]
```diff
@@ -270,6 +271,53 @@ private void testAvroSchemaEvolution()
         }
     }
 
+    @Test(dataProvider = "storage_formats", groups = SMOKE)
+    public void testReadUnionAsStruct(String storageFormat)
```
Comment (phd3) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestReadUniontype.java:275]: Can we merge this into existing testUnionTypeSchemaEvolution ? Or atleast call this something explicit like testUnionToStructSchemaEvolution

**Comment by phd3** [comment]
cc @electrum @kokosing

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSplitManager.java [hunk 2]
```diff
@@ -410,6 +411,14 @@ private TableToPartitionMapping getTableToPartitionMapping(ConnectorSession sess
         for (int i = 0; i < min(partitionColumns.size(), tableColumns.size()); i++) {
             HiveType tableType = tableColumns.get(i).getType();
             HiveType partitionType = partitionColumns.get(i).getType();
+
+            if (tableType.getCategory() == ObjectInspector.Category.UNION) {
+                tableType = unionToStruct(tableType, typeManager);
```
[the remark this answers is not in the document]
Comment (leetcode-1533) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSplitManager.java:416]: The change is bigger than it looks:

Instead of changing the "canCoerce" method, which is used by the coordinator when sending the split. I early casted the data type for the TableToPartitionMapping(part of the Split's information)'s columnCoercions field. 
Therefore, when worker node inspect the split's partition information, start to read the split information assigned by the coordinator, workers already found the partition type is struct instead of union.(This is used when creating the pagesource)

However, casting when creating StructCoercer is still needed, if we think this is a good approach, I can double check and see if creating StructCoercer can use the information from the TableToPartitionMapping

[the remark this answers is not in the document]
Comment (leetcode-1533) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSplitManager.java:416]: First of all:
Since union type only allows one type at a time, I am not sure dereferencing for the union column is supported in general. I tried to look for examples as how union type are used, most examples I saw select the whole union column. Can you find a example that is doing dereferencing for the union type? I want to know what subscript is being used..


I don't think eventually "everything is struct - when they come to the worker." is possible, since union is a component type in ORC:
`Compound types have children columns that hold the values for their sub-elements. For example, a struct column has one child column for each field of the struct. Lists always have a single child column for the element values and maps always have two child columns. Union columns have one child column for each of the variant`

ORC reader/writer need to know it is handling a union column.

Do you have a preference for implementing "getSupportedType" or keep casting and using isSupportedType? seems the change is small, to reduce scope we should continue use isSupportedType?

[the remark this answers is not in the document]
Comment (leetcode-1533) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSplitManager.java:416]: union to struct cast is only casting we have when building TableToPartitionMapping

[the remark this answers is not in the document]
Comment (phd3) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSplitManager.java:416]: > Since union type only allows one type at a time, I am not sure dereferencing for the union column is supported in general.

Ultimately to process union types - I'd expect the end user to check the tag and get the non-null value somehow. Or expose the N values as separate column. i.e. the user needs to look at what's inside for processing. And if this is something that only throws an error if pushdown happens but not otherwise - then that's a bug. 

> I don't think eventually "everything is struct - when they come to the worker."

I just meant this in terms of things around hive pagesource code handling table and partitions metadata. For ORC, we have a layer on top which is the UnionColumnReader.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveCoercionPolicy.java [hunk 3]
```diff
@@ -123,11 +110,8 @@ private boolean canCoerceForList(HiveType fromHiveType, HiveType toHiveType)
         return fromElementType.equals(toElementType) || canCoerce(fromElementType, toElementType);
     }
 
-    private boolean canCoerceForStruct(HiveType fromHiveType, HiveType toHiveType)
+    private boolean canCoerceForStructOrUnion(HiveType fromHiveType, HiveType toHiveType)
     {
-        if (fromHiveType.getCategory() != Category.STRUCT || toHiveType.getCategory() != Category.STRUCT) {
-            return false;
-        }
         List<String> fromFieldNames = ((StructTypeInfo) fromHiveType.getTypeInfo()).getAllStructFieldNames();
```
[the remark this answers is not in the document]
Comment (leetcode-1533) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveCoercionPolicy.java:115]: yeah， my bad, adding back.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePageSource.java [hunk 4]
```diff
@@ -354,12 +354,16 @@ private static Optional<Function<Block, Block>> createCoercer(TypeManager typeMa
             return Optional.of(new MapCoercer(typeManager, fromHiveType, toHiveType));
         }
         if (isRowType(fromType) && isRowType(toType)) {
-            if (fromHiveType.getCategory() == ObjectInspector.Category.UNION || toHiveType.getCategory() == ObjectInspector.Category.UNION) {
-                HiveType fromHiveTypeStruct = HiveType.toHiveType(fromType);
-                HiveType toHiveTypeStruct = HiveType.toHiveType(toType);
-                return Optional.of(new StructCoercer(typeManager, fromHiveTypeStruct, toHiveTypeStruct));
+            HiveType fromHiveTypeStruct = fromHiveType;
```
[the remark this answers is not in the document]
Comment (leetcode-1533) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePageSource.java:357]: let me double check if HiveType.toHiveType() is idempotent .

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSplitManager.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (leetcode-1533) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSplitManager.java:416]: summary from our discussion: remove this casting and make sure the rest still works.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestReadUniontype.java [hunk 5]
```diff
@@ -270,6 +271,53 @@ private void testAvroSchemaEvolution()
...
+
+        onHive().executeQuery(format(
+                "create table %s(" +
+                        "c1 uniontype<struct<a:string,b:string>, struct<c:string,d:string>>) " +
+                        "partitioned by (p int) STORED AS %s",
+                TABLE_READ_UNION_AS_STRUCT,
+                storageFormat));
+        switch (storageFormat) {
+            case "AVRO":
+                // todo:  test
```
[the remark this answers is not in the document]
Comment (leetcode-1533) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestReadUniontype.java:294]: let's make sure the ORC flow ent to end is satisfying and I will move on to AVRO

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSplitManager.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (leetcode-1533) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSplitManager.java:416]: Now I understand the reason why I previously implemented in this way:

It is due to supporting dereference for union type. Especially for functions getHiveTypeForDereferences() and getHiveDereferenceNames(). Previously, Trino do not support dereference for union type, as we can see from these two functions, there were checkArgument statement enforcing the hiveType being struct type. (If you know how Hive has a dereference for union type, please let me know)

However, since right now we are implementing read union as struct, naturally, this read should be support dereference. 

So we have 3 solutions:

1. disable dereference for union type, making sure when read union type, the HiveColumnProjectionInfo is always empty
  needs further investigation.

2. casting union to struct while dereferencing: 
this needs major interfaces changes for getHiveTypeForDereferences() and getHiveDereferenceNames() and along its calling chain. In order to pass in typeManagers, which is necessary for casting union to struct. 

3. early casting union to struct, and using its struct representation for dereferencing. 
This implementation. 

I don't like approach 2, since it changes too much for this corner case. For this corner case, our aim is to reduce impact radius and ensure correctness rather than efficiency.

[the remark this answers is not in the document]
Comment (leetcode-1533) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSplitManager.java:416]: moving on, we should be clear that Trino's Hive library only supports dereference for struct type(for Orc, Parquet, and Avro).

[the remark this answers is not in the document]
Comment (leetcode-1533) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSplitManager.java]: Summary: we concluded to take approach 4.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestReadUniontype.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (leetcode-1533) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestReadUniontype.java]: we won't support AVRO

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestReadUniontype.java [hunk 6]
```diff
@@ -270,6 +271,53 @@ private void testAvroSchemaEvolution()
...
+    private void testORCReadUnionAsStruct(String tableName) {
+        onHive().executeQuery(format("insert into table %s partition(p) " +
+                        "SELECT create_union(1, named_struct('a', 'a1', 'b', 'b1'), named_struct('c', 'ignores', 'd', 'ignore')), 999 from (select 1) t",
+                tableName));
+
+        onHive().executeQuery(format("ALTER TABLE %s CHANGE COLUMN c1 c1 " +
+                        " STRUCT<tag:int, field0:STRUCT<a:STRING, b:STRING>, field1:STRUCT<c:STRING, d:STRING>>",
+                tableName));
+
+        onHive().executeQuery(format("insert into table %s partition(p) " +
```
[the remark this answers is not in the document]
Comment (leetcode-1533) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestReadUniontype.java]: DONE

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestReadUniontype.java (hunk 1 above)
Comment (leetcode-1533) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestReadUniontype.java]: done.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSplitManager.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (leetcode-1533) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSplitManager.java]: And we won't be supporting this feature for AVRO type, due to Hive Avro's implementation using Table schema for deserialization.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePageSource.java [hunk 7]
```diff
@@ -354,12 +354,10 @@ private static Optional<Function<Block, Block>> createCoercer(TypeManager typeMa
             return Optional.of(new MapCoercer(typeManager, fromHiveType, toHiveType));
         }
         if (isRowType(fromType) && isRowType(toType)) {
-            if (fromHiveType.getCategory() == ObjectInspector.Category.UNION || toHiveType.getCategory() == ObjectInspector.Category.UNION) {
-                HiveType fromHiveTypeStruct = HiveType.toHiveType(fromType);
-                HiveType toHiveTypeStruct = HiveType.toHiveType(toType);
-                return Optional.of(new StructCoercer(typeManager, fromHiveTypeStruct, toHiveTypeStruct));
-            }
-            return Optional.of(new StructCoercer(typeManager, fromHiveType, toHiveType));
+            HiveType fromHiveTypeStruct = (fromHiveType.getCategory() == ObjectInspector.Category.UNION) ? HiveType.toHiveType(fromType) : fromHiveType;
```
Comment (phd3) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePageSource.java:357]: can we make this a separate commit ? This is an improvement - but not required for the fix.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveType.java [hunk 8]
```diff
@@ -17,7 +17,6 @@
 import com.fasterxml.jackson.annotation.JsonValue;
 import com.google.common.collect.ImmutableList;
 import io.trino.plugin.hive.metastore.StorageFormat;
-import io.trino.spi.type.TinyintType;
```
Comment (phd3) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveType.java]: unrelated ?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveCoercionPolicy.java [hunk 9]
```diff
@@ -109,29 +98,33 @@ private boolean canCoerceForMap(HiveType fromHiveType, HiveType toHiveType)
...
         HiveType toElementType = HiveType.valueOf(((ListTypeInfo) toHiveType.getTypeInfo()).getListElementTypeInfo().getTypeName());
-        return fromElementType.equals(toElementType) || canCoerce(fromElementType, toElementType);
+        return fromElementType.equals(toElementType) || canCoerce(fromElementType, toElementType, hiveTimestampPrecision);
     }
 
-    private boolean canCoerceForStruct(HiveType fromHiveType, HiveType toHiveType)
+    private boolean canCoerceForStructOrUnion(HiveType fromHiveType, HiveType toHiveType, HiveTimestampPrecision hiveTimestampPrecision)
     {
-        if (fromHiveType.getCategory() != Category.STRUCT || toHiveType.getCategory() != Category.STRUCT) {
+        if ((fromHiveType.getCategory() != Category.STRUCT && fromHiveType.getCategory() != Category.UNION)
```
Comment (phd3) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveCoercionPolicy.java:117]: nit: may be abstract this out in `isStructOrUnion`

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveCoercionPolicy.java [hunk 10]
```diff
@@ -109,29 +98,33 @@ private boolean canCoerceForMap(HiveType fromHiveType, HiveType toHiveType)
...
+                || (toHiveType.getCategory() != Category.STRUCT && toHiveType.getCategory() != Category.UNION)) {
             return false;
         }
-        List<String> fromFieldNames = ((StructTypeInfo) fromHiveType.getTypeInfo()).getAllStructFieldNames();
-        List<String> toFieldNames = ((StructTypeInfo) toHiveType.getTypeInfo()).getAllStructFieldNames();
-        List<HiveType> fromFieldTypes = extractStructFieldTypes(fromHiveType);
-        List<HiveType> toFieldTypes = extractStructFieldTypes(toHiveType);
+        HiveType fromHiveTypeStruct = (fromHiveType.getCategory() == ObjectInspector.Category.UNION) ? HiveType.toHiveType(fromHiveType.getType(typeManager, hiveTimestampPrecision)) : fromHiveType;
```
Comment (phd3) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveCoercionPolicy.java:121]: nit: since this is invoked multiple times - may be have separate getStructFromUnion

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePageSource.java (hunk 7 above)
Comment (leetcode-1533) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePageSource.java:357]: ok

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSplitManager.java [hunk 11]
```diff
@@ -456,6 +473,13 @@ private TableToPartitionMapping getTableToPartitionMappingByColumnNames(SchemaTa
             tableToPartitionColumns.put(tableColumnIndex, partitionColumnIndex);
             Column partitionColumn = partitionColumns.get(partitionColumnIndex);
             HiveType partitionType = partitionColumn.getType();
+
+            if (tableType.getCategory() == ObjectInspector.Category.UNION) {
```
[the remark this answers is not in the document]
Comment (leetcode-1533) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSplitManager.java]: this is outdated..

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveCoercionPolicy.java (hunk 10 above)
Comment (leetcode-1533) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveCoercionPolicy.java:121]: Suggest name as convertUnionToStruct

Comment (leetcode-1533) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveCoercionPolicy.java:121]: getStructFromUnion is a name seems like union is a collections of properties and getting one of the struct property from the union.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePageSource.java (hunk 7 above)
Comment (leetcode-1533) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePageSource.java:357]: call it " HivePageSource coercer only convert to struct when hivetype is union" please suggest a better name

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSplitManager.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (leetcode-1533) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSplitManager.java]: if you are asking  HiveType.toHiveType(fromType), nested is taken care of, basically, it convert the Trino's representation(row) to hive struct.

**Comment by leetcode-1533** [comment]
hi, I updated the PR, it is ready to review. Rebase on latest master branch, which include changes from: https://github.com/trinodb/trino/pull/15278

**Review by phd3 — APPROVED** [review]

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveCoercionPolicy.java [hunk 12]
```diff
@@ -109,29 +100,32 @@ private boolean canCoerceForMap(HiveType fromHiveType, HiveType toHiveType)
...
-        if (fromHiveType.getCategory() != Category.STRUCT || toHiveType.getCategory() != Category.STRUCT) {
+        if (!isStructOrUnion(fromHiveType) || !isStructOrUnion(toHiveType)) {
             return false;
         }
-        List<String> fromFieldNames = ((StructTypeInfo) fromHiveType.getTypeInfo()).getAllStructFieldNames();
-        List<String> toFieldNames = ((StructTypeInfo) toHiveType.getTypeInfo()).getAllStructFieldNames();
-        List<HiveType> fromFieldTypes = extractStructFieldTypes(fromHiveType);
-        List<HiveType> toFieldTypes = extractStructFieldTypes(toHiveType);
+        HiveType fromHiveTypeStruct = (fromHiveType.getCategory() == ObjectInspector.Category.UNION) ? convertUnionToStruct(fromHiveType, typeManager, hiveTimestampPrecision) : fromHiveType;
+        HiveType toHiveTypeStruct = (toHiveType.getCategory() == ObjectInspector.Category.UNION) ? convertUnionToStruct(toHiveType, typeManager, hiveTimestampPrecision) : toHiveType;
```
[the remark this answers is not in the document]
Comment (leetcode-1533) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveCoercionPolicy.java:123]: done.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestReadUniontype.java [hunk 13]
```diff
@@ -268,6 +268,77 @@ public void testUnionTypeSchemaEvolution(String storageFormat)
         }
     }
 
+    @Test(groups = SMOKE)
+    public void testORCUnionToStructSchemaEvolution()
```
[the remark this answers is not in the document]
Comment (leetcode-1533) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestReadUniontype.java:272]: Please let me know if it looks good to you.

    /**
     * When reading AVRO file, Trino needs the schema information from Hive metastore to deserialize Avro files.
     * Therefore, when an ALTER table was issued in which the hive metastore changed the schema into an incompatible format,
     * from Union to Struct or from Struct to Union in this case, Trino could not read those Avro files using the modified Hive metastore schema.
     * However, when reading ORC files, Trino does not need schema information from Hive metastore to deserialize ORC files.
     * Therefore, it can read ORC files even after changing the schema.
     */

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestReadUniontype.java [hunk 14]
```diff
@@ -268,6 +268,77 @@ public void testUnionTypeSchemaEvolution(String storageFormat)
...
+        onHive().executeQuery(format("ALTER TABLE %s CHANGE COLUMN c1 c1 " +
+                        " STRUCT<tag:int, field0:STRUCT<a:STRING, b:STRING>, field1:STRUCT<c:STRING, d:STRING>>",
+                tableReadUnionAsStruct));
+
+        onHive().executeQuery(format("INSERT INTO TABLE %s partition(p) " +
+                        "SELECT named_struct('tag', 0, 'field0', named_struct('a', 'a11', 'b', 'b1b'), 'field1',named_struct('c', 'ignores', 'd', 'ignores')), 100 from (select 1) t",
+                tableReadUnionAsStruct));
+        // using dereference
+        QueryResult selectAllResult = onTrino().executeQuery(format("SELECT c1.field0 FROM hive.default.%s", tableReadUnionAsStruct));
+        assertEquals(selectAllResult.rows().size(), 2);
```
[the remark this answers is not in the document]
Comment (leetcode-1533) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestReadUniontype.java]: done.

**Comment by leetcode-1533** [comment]
Hi, it has been updated, please take a further look.

**Review by phd3 — APPROVED** [review]
